### PR TITLE
Management API implementation for Required Settings

### PR DESCRIPTION
**Pull request by peterojo**
**Description**

Add the management API into the required settings of Magento.
Create auto configuration option for new merchants (without saved configuration), which uses management API to fetch merchant accounts, live prefixURLs, client keys, and generates new webhook credentials.

## Important Info
This update drops support for magento 2.3.* and php 7.3 since we expect to release this with the next major version, and [M2.3.7 reached end of support in 04-22](https://devdocs.magento.com/guides/v2.3/release-notes/open-source-2-3-7.html)

**Tested scenarios**

- New merchant configuration in live
- New merchant configuration in test
- Existing merchant updates on live/test

### CODE REVIEW

File: Controller/Adminhtml/Configuration/WebhookTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    public function __construct(
+        Context $context,
+        ManagementHelper $managementApiHelper,
+        \Adyen\Payment\Helper\Data $adyenHelper,
+        JsonFactory $resultJsonFactory
+    ) {
+        parent::__construct($context);
+        $this->managementApiHelper = $managementApiHelper;
+        $this->adyenHelper =$adyenHelper;
```
Comment (Morerice) [inline Controller/Adminhtml/Configuration/WebhookTest.php:57]:
```suggestion
        $this->adyenHelper = $adyenHelper;
```

File: Controller/Adminhtml/Configuration/WebhookTest.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+{
+    /**
+     * @var ManagementHelper
+     */
+    private $managementApiHelper;
+
+    /**
+     * @var \Adyen\Payment\Helper\Data
+     */
+    protected $adyenHelper;
```
Comment (Morerice) [inline Controller/Adminhtml/Configuration/WebhookTest.php:42]:
```suggestion
    protected $dataHelper;
```

File: Controller/Adminhtml/Configuration/WebhookTest.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    ) {
+        parent::__construct($context);
+        $this->managementApiHelper = $managementApiHelper;
+        $this->adyenHelper =$adyenHelper;
+        $this->resultJsonFactory = $resultJsonFactory;
+    }
+
+    /**
+     * @return \Magento\Framework\App\ResponseInterface|\Magento\Framework\Controller\ResultInterface|mixed|string
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
```
Comment (Morerice) [inline Controller/Adminhtml/Configuration/WebhookTest.php:63]: Can we import the namespaces and add a return type to the function?

File: Helper/BaseUrlHelper.php [hunk 4]
```diff
@@ -50,16 +50,28 @@ public function __construct(UrlInterface $url, State $state, BackendDataHelper $
      * @param null|int|string $storeId
      * @return string
      */
-    public function getStoreBaseUrl($storeId)
+    public function getStoreBaseUrl($storeId, $ignoreAdmin = false)
```
Comment (Morerice) [inline Helper/BaseUrlHelper.php:53]: Can we add a return type?

File: Helper/Config.php [hunk 5]
```diff
@@ -51,18 +51,48 @@ class Config
...
         $this->encryptor = $encryptor;
+        $this->configWriter = $configWriter;
+    }
+
+    /**
+     * @param $mode
+     * @param null $storeId
+     * @return string
+     */
+    public function getApiKey($mode, $storeId = null)
```
Comment (Morerice) [inline Helper/Config.php:81]: Can we add a return type?

File: Helper/Config.php [hunk 6]
```diff
@@ -51,18 +51,48 @@ class Config
...
+
+        return $this->encryptor->decrypt($apiKey);
+    }
+
+    /**
+     * @param $mode
+     * @param $storeId
+     * @return bool|mixed
+     */
+    public function getClientKey($mode, $storeId = null)
```
Comment (Morerice) [inline Helper/Config.php:93]: Can we add a return type?

File: Helper/Config.php [hunk 7]
```diff
@@ -110,47 +140,33 @@ public function getNotificationsPassword($storeId = null)
...
     /**
-     * Retrieve flag for notifications_can_cancel
-     *
-     * @param int $storeId
-     * @return bool
+     * @param int|null $storeId
+     * @return string
      */
-    public function getNotificationsCanCancel($storeId = null)
+    public function getWebhookUrl($storeId = null)
```
Comment (Morerice) [inline Helper/Config.php:146]: Can we add a return type and update the parameter type?

File: Helper/Config.php [hunk 8]
```diff
@@ -110,47 +140,33 @@ public function getNotificationsPassword($storeId = null)
...
 
     /**
-     * Retrieve flag for notifications_ip_check
+     * Retrieve flag for notifications_can_cancel
      *
      * @param int $storeId
      * @return bool
      */
-    public function getNotificationsIpCheck($storeId = null)
+    public function getNotificationsCanCancel($storeId = null)
```
Comment (Morerice) [inline Helper/Config.php:166]: Can we add a return type and update the parameter type?

File: Helper/Data.php [hunk 9]
```diff
@@ -1484,18 +1485,22 @@ public function formatTerminalAPIReceipt($paymentReceipt)
      *
      * @param null|int|string $storeId
      * @param string|null $apiKey
+     * @param bool|null $demoMode
      * @return \Adyen\Client
      * @throws \Adyen\AdyenException
+     * @throws \Magento\Framework\Exception\NoSuchEntityException
      */
-    public function initializeAdyenClient($storeId = null, $apiKey = null)
+    public function initializeAdyenClient($storeId = null, $apiKey = null, ?bool $demoMode = null)
```
Comment (Morerice) [inline Helper/Data.php:1493]: Can we add a return type?

File: Helper/ManagementHelper.php [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Store\Model\StoreManager;
+use Adyen\Payment\Logger\AdyenLogger;
+
+class ManagementHelper
+{
+    /**
+     * @var Data
+     */
+    private $adyenHelper;
```
Comment (Morerice) [inline Helper/ManagementHelper.php:29]:
```suggestion
    private $dataHelper;
```

File: Helper/ManagementHelper.php [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+     * @param string $apiKey
+     * @param string $merchantId
+     * @param string $username
+     * @param string $password
+     * @param string $url
+     * @param bool $demoMode
+     * @throws AdyenException
+     * @throws NoSuchEntityException
+     */
+    public function setupWebhookCredentials(
```
Comment (Morerice) [inline Helper/ManagementHelper.php:118]: Can we add a return type?

File: Helper/ManagementHelper.php [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+        $response = $management->merchantWebhooks->generateHmac($merchantId, $webhookId);
+        $hmac = $response['hmacKey'];
+        $mode = $demoMode ? 'test' : 'live';
+        $this->configHelper->setConfigData($hmac, 'notification_hmac_key_' . $mode, Config::XML_ADYEN_ABSTRACT_PREFIX);
+    }
+
+    /**
+     * @throws AdyenException|NoSuchEntityException
+     */
+    public function getAllowedOrigins($apiKey, $environment)
```
Comment (Morerice) [inline Helper/ManagementHelper.php:180]: Can we add a return type and update the parameter type?

File: Helper/ManagementHelper.php [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+        $response = $management->allowedOrigins->list();
+
+        return !empty($response) ? array_column($response['data'], 'domain') : [];
+    }
+
+    /**
+     * @throws AdyenException|NoSuchEntityException
+     */
+    public function saveAllowedOrigin($apiKey, $environment, $domain)
```
Comment (Morerice) [inline Helper/ManagementHelper.php:192]: Can we add a return type and update the parameter type?

File: Helper/ManagementHelper.php [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+    {
+        $management = $this->getManagementApiService($apiKey, $environment);
+
+        $management->allowedOrigins->create(['domain' => $domain]);
+    }
+
+    /**
+     * @throws AdyenException|NoSuchEntityException
+     */
+    private function getManagementApiService($apiKey, $environment): Management
```
Comment (Morerice) [inline Helper/ManagementHelper.php:202]: Can we update the parameter type?

File: Helper/ManagementHelper.php [hunk 15]
```diff
@@ -0,0 +1,239 @@
...
+
+        return new Management($client);
+    }
+
+    /**
+     * @param string $merchantId
+     * @return mixed|string
+     * @throws NoSuchEntityException
+     */
+    public function webhookTest(string $merchantId)
```
Comment (Morerice) [inline Helper/ManagementHelper.php:219]: Can we update the parameter type?

File: Model/Config/Adminhtml/ConfigurationWizard.php [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+            [
+                'id' => 'adyen_configuration_action_reset',
+                'label' => __('Reconfigure'),
+            ]
+        );
+
+        return $button->toHtml();
+    }
+
+    public function getMerchantAccountsUrl()
```
Comment (Morerice) [inline Model/Config/Adminhtml/ConfigurationWizard.php:71]: Can we add a return type?

File: Model/Config/Adminhtml/ConfigurationWizard.php [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return $this->getUrl('adyen/configuration/merchantaccounts');
+    }
+
+    public function getStoreId()
```
Comment (Morerice) [inline Model/Config/Adminhtml/ConfigurationWizard.php:76]: Can we add a return type?

File: Model/Config/Adminhtml/ConfigurationWizard.php [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return $this->_storeManager->getStore()->getId();
+    }
+
+    public function testConfigured() {
```
Comment (Morerice) [inline Model/Config/Adminhtml/ConfigurationWizard.php:81]: Can we add a return type?

File: Model/Config/Adminhtml/ConfigurationWizard.php [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+        $merchantAccount = boolval($this->configHelper->getMerchantAccount());
+        $clientKeyTest = boolval($this->configHelper->getClientKey('test', $this->getStoreId()));
+        $notificationUsername = boolval($this->configHelper->getNotificationsUsername($this->getStoreId()));
+        $notificationPassword = boolval($this->configHelper->getNotificationsPassword($this->getStoreId()));
+
+        return $merchantAccount || $clientKeyTest || $notificationUsername || $notificationPassword;
+    }
+
+    public function liveConfigured() {
```
Comment (Morerice) [inline Model/Config/Adminhtml/ConfigurationWizard.php:90]: Can we add a return type?

File: Model/Config/Adminhtml/MerchantAccounts.php [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Model\Config\Adminhtml;
+
+class MerchantAccounts extends \Magento\Config\Block\System\Config\Form\Field
+{
+    protected $_template = 'Adyen_Payment::config/required_settings.phtml';
+
+    public function render(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Comment (Morerice) [inline Model/Config/Adminhtml/MerchantAccounts.php:31]: Can we add a return type in the functions of this class?

File: Model/Config/Adminhtml/WebhookTest.php [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Model\Config\Adminhtml;
+
+class WebhookTest extends \Magento\Config\Block\System\Config\Form\Field
+{
+    protected $_template = 'Adyen_Payment::config/webhook_test.phtml';
+
+    public function render(\Magento\Framework\Data\Form\Element\AbstractElement $element)
```
Comment (Morerice) [inline Model/Config/Adminhtml/WebhookTest.php:31]: Can we add a return type in the functions of this class?

File: Model/Config/Source/ConfigurationModes.php [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+namespace Adyen\Payment\Model\Config\Source;
+
+use Magento\Framework\Option\ArrayInterface;
+
+class ConfigurationModes implements ArrayInterface
+{
+    /**
+     * @return array
+     */
+    public function toOptionArray()
```
Comment (Morerice) [inline Model/Config/Source/ConfigurationModes.php:20]: Can we add a return type in the functions of this class?

File: Model/Config/Source/MerchantAccounts.php [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+ */
+
+namespace Adyen\Payment\Model\Config\Source;
+
+class MerchantAccounts implements \Magento\Framework\Data\OptionSourceInterface
+{
+    /**
+     * @var \Adyen\Payment\Helper\Data
+     */
+    protected $_adyenHelper;
```
Comment (Morerice) [inline Model/Config/Source/MerchantAccounts.php:32]:
```suggestion
    protected $dataHelper;
```

File: Controller/Adminhtml/Configuration/WebhookTest.php [hunk 24]
```diff
@@ -0,0 +1,78 @@
+<?php
+/**
+ *                       ######
```
Comment (Morerice) [inline Controller/Adminhtml/Configuration/WebhookTest.php:3]: Can we remove the bulky header from all the new files?

**Review by Morerice — CHANGES_REQUESTED** [review]

File: Controller/Adminhtml/Configuration/WebhookTest.php (hunk 3 above)
Comment (peterojo) [reply Controller/Adminhtml/Configuration/WebhookTest.php:63]: interface definition does not declare a return type so we have to leave it out here too

File: Helper/Config.php (hunk 6 above)
Comment (peterojo) [reply Helper/Config.php:93]: return type here is indeterminate

File: Model/Config/Adminhtml/MerchantAccounts.php (hunk 20 above)
Comment (peterojo) [reply Model/Config/Adminhtml/MerchantAccounts.php:31]: interface doesn't declare a return type

**Review by Morerice — APPROVED** [review]

**Review by Morerice — APPROVED** [review]

File: Helper/ManagementHelper.php [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+        $management = $this->getManagementApiService($apiKey, $demoMode ? 'test' : 'live');
+        $merchantAccounts = [];
+        $page = 1;
+        $pageSize = 100;
+        //get the merchant accounts using get /merchants.
+        $responseMerchants = $management->merchantAccount->list(["pageSize" => $pageSize]);
+        while (count($merchantAccounts) < $responseMerchants['itemsTotal']) {
+            foreach ($responseMerchants['data'] as $merchantAccount) {
+                $defaultDC = array_filter($merchantAccount['dataCenters'], function ($dc) {
+                    return $dc['name'] = 'default';
```
Comment (michaelpaul) [inline Helper/ManagementHelper.php:82]:
```suggestion
                    return $dc['name'] === 'default';
```

**Review by michaelpaul — CHANGES_REQUESTED** [review]

File: Controller/Process/Json.php [hunk 26]
```diff
@@ -252,16 +252,15 @@ private function processNotification(array $response, $notificationMode)
         }
 
         // Validate if Ip check is enabled and if the notification comes from a verified IP
-        if ($this->configHelper->getNotificationsIpCheck() && !$this->isIpValid()) {
+        if (!$this->isIpValid()) {
```
Comment (acampos1916) [inline Controller/Process/Json.php:255]: Let's remember to mention in the release notes that the IP and HMAC checks will always be on

File: Model/Config/Adminhtml/WebhookTest.php [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    public function getButtonHtml()
+    {
+        $button = $this->getLayout()->createBlock(
+            'Magento\Backend\Block\Widget\Button'
+        )->setData(
+            [
+                'id' => 'adyen_webhook_test',
+                'label' => __('Test webhook url'),
```
Comment (acampos1916) [inline Model/Config/Adminhtml/WebhookTest.php:44]:
```suggestion
                'label' => __('Test Webhook URL'),
```

**Review by acampos1916 — DISMISSED** [review]

**Review by RokPopov — APPROVED** [review]

**Review by acampos1916 — APPROVED** [review]

**Review by RokPopov — APPROVED** [review]

**Pull request merged**